torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1185,14 +1434,73 @@ def _register_param_handle(self, handle: FlatParamHandle) -> None:
         if handle not in self._handles:
             self._handles.append(handle)
 
+    @torch.no_grad()
+    def _unshard(
+        self,
+        handles: List[FlatParamHandle],
+    ) -> None:
+        """
+        Unshards the handles in ``handles``. If the handles are in
+        :meth:`summon_full_params` and are using mixed precision, then they are
+        forced to full precision.
+
+        Postcondition: Each handle's ``FlatParameter`` 's data is the padded
+        unsharded flattened parameter on the compute device.
+        """
+        with torch.cuda.stream(self._streams["all_gather"]):
+            for handle in handles:
+                handle.pre_unshard()
```

Comment:
IIUC, `pre_unshard` might trigger an H2D copy which uses PCIe and can overlap with allgather (and compute). Do we need to move this to a dedicated stream so enable overlapping?

Reply:
The existing code had a mixed-precision stream just for allocating the low precision shard. I removed it because I did not think it was achieving any overlapping. However, I think I may have been mistaken in that, as you said, we can start H2D copies early and overlap.

I think the correct thing to do is to have a "pre-unshard" stream (or a "pre-all-gather" stream, naming is flexible) for these H2D copies. The existing mixed-precision stream may have been a slight misnomer. I will add this stream back for `pre_unshard()`.

Reply:
I thought non-blocking H2D possibly just makes overlapping work without separated stream..?

Reply:
I will share a learning note tomorrow in our working group to give some explanation for this.